### PR TITLE
Create algorithm to find all possible meeting times

### PR DESCRIPTION
**Pull request by shreyachatterjee00**
Passes all pre-written tests: 
* Functionality for finding a meeting time given a meeting request and list of events on that day 
* Works for conflicting events, nested events, no events, etc

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java
Comment (megankj) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java:37]: This file + style.css look unrelated to this PR.  Was this an unintentional commit?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Please add a javadoc comment explaining what this method does.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Lines for java files should not be longer than 100 characters.  Multi-line comments should be in this form:

```
  /**
   * multi line comment that is shorter than 100 characters
   * that goes on to the next line
   */
```

Please update here and throughout.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Is this the comment for the entire method?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:105]: Don't orphan else statements like this, it might be hard to read!
```suggestion
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: 30 or 10?  Also it's better to make these constants so they're easier to update and easier to read.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion
        currTime = longestMeetingEnd;
        startTime = longestMeetingEnd;
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion
    if (currTime > TimeRange.END_OF_DAY && currDuration >= mtngDuration) {
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: does it end at END_OF_DAY because TimeRange can't go into the next day?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:109]: This is looking good, but it's pretty long for one method so I wonder if you can break this up into more logical helper methods.  From the looks of it, some of the steps you're doing are:
1) setting up the map of event times w/ conflicts
2) processing the map to determine where the "free time" is in the map 

Maybe at least those can be broken up?  Smaller helper methods will also help document the code.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I may be wrong, but it looks like you only ever care about the longest meeting for a given start time.  Do you need to store the other ones?  Could that be pre-processed?  (Or does that increase complexity?)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: To make it clear what this boolean is, I suggest adding a comment like this:
```suggestion
            meetingTimes.add(TimeRange.fromStartEnd(startTime, currTime, /* inclusive= */ false));
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion
        currTime += 10;
```

**Review by megankj — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java
Comment (shreyachatterjee00) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java:37]: So I had just merged multiple other PRs into master, and then pulled those before starting development on this branch, so I thought I'd make the few design changes that emerged from merging those PRs, on this branch. I thought it wouldn't be worth it to make a whole new branch for a couple lines of code, but I probably just should have because yeah, this has nothing to do with the rest of the PR.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (shreyachatterjee00) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Yeah, they provided START_OF_DAY and END_OF_DAY constants which are what the meetings are supposed to fall within.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java
Comment (megankj) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java:37]: alright, that's fine. thanks for the explanation!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: use javadoc formatting {@code TimeRange} which will help link all these objects together when looking at the codebase in github!

Comment (megankj) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: also this javadoc should probably go above the query method, not the entire class

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I think it would be good to explain what mandatory means

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I don't think you actually need to iterate through every single attendee, right?  You just need to determine if this event contains any of the provided attendees, and if so, add the event time to your map.

You may be able to remove some of this nesting by using [java streams anyMatch](https://www.geeksforgeeks.org/stream-anymatch-java-examples/) to find out of this event is potentially conflicting with the request.  anyMatch is nice because it's short-circuiting and will terminate if it finds a match, not having to go through all attendees.

boolean eventHasAttendee = attendees.stream().anyMatch(attendee -> event.getAttendees().contains(attendee));

if (eventHasAttendee) {
  addMandatoryEvent(event);
}

And then your addMandatory event function can add the event to your map (unless it's shorter than the existing entry).

(As a side note, code with lots of nesting has the potential to introduce bugs since it makes it hard to follow the conditional branching that may or may not occur)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: Method comments should be in the form "Creates and returns.." instead of "Create and return"

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: make this comment shorter than 100 characters so doesn't wrap the line like this

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Does need to be declared here?  It's immediately re-assigned on line 97.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Methods with names like isFoo are expected to return booleans.  Instead, since this returns an int, it should be something like getLastEventEnd or something that better describes what int you're getting here.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I mentioned this in our 1:1, but it's better to keep the code a little closer together like this:
```suggestion
      } else {
      // If there is no event at this time, add MIN_INCREMENT minutes to the duration and current time, and run the loop again. 
```

Please update here and elsewhere.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: use MIN_INCREMENT?

**Review by megankj — CHANGES_REQUESTED** [review]
looking good, thanks for making the changes!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (shreyachatterjee00) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I  need to keep track of the current longest meeting end because I'm comparing against it in the isConflict function, but I can call this variable something else to make it clearer. Instead of making this a variable at all I could also just pass "mandatoryEvents.get(currTime)" into the function directly, but I feel like that's not as clean

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:35]: can you reformat this comment so it's less than 100 chars per line?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion
    // Mandatory events are events that must be attended by an attendee in the {@code MeetingRequest}.
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: is there a reason you make a copy of this?

Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: this is a small nit, but none of these need to be local variables, you can just inline them in the makeMandatoryEventMap and findMeetingTimes methods below

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (megankj) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion
  * Creates and returns an ArrayList of {@code TimeRange} objects that span available meeting times.
```

**Review by megankj — CHANGES_REQUESTED** [review]
Does running "make pretty" format java files?  I'm wondering if it will auto-truncate lines to 100 characters for you.

@tjgoog feel free to also leave comments at this point.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (tjgoog) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: To get technical here... this is a new reference pointing to an existing object, so if you do want to make a copy that you can modify without affecting the original (which is a good practice if you are modifying input data), you would do something like
`LinkedList<Event> modifiableEvents = new LinkedList<>(events);`

If you don't make a copy, this is effectively creating an alias for `events`, so if you prefer a different name, just change the parameter name at the top of the method.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Document that if there are two events at the same start time, the end time stored in the map will be the latest end time for that start time.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:55]: Yay Streams! I like them more than I probably should. (Just a comment, not requested change)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Since the javadoc is saying that this will return an ArrayList, should change method declaration to be ArrayList<TimeRange> or change doc to be Collection.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (tjgoog) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Even better: Use `{@link TimeRange}` so that relevant tools will make this a link to the TimeRange class definition. Also, `{@link }`-ify `ArrayList`.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion
      tempTime += MIN_INCREMENT;
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: When execution gets here, `currTime` could be (will be, assuming `END_OF_DAY` is a multiple of `MIN_INCREMENT`, I think) == `END_OF_DAY`. So shouldn't the first comparison be `>=`?

**Review by tjgoog — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (shreyachatterjee00) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: yeah, should be

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion
    // Mandatory events are events that must be attended by a mandatory attendee in the {@link Meeting Request}
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java
Comment (tjgoog) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Because you're not modifying anything in `events`, no need to create a copy of it, so I think this should just be removed and use `events` in the `for` loop.

**Review by tjgoog — APPROVED** [review]
Approved! Please address my small comments and get Megan's approval too before merging.

**Review by megankj — APPROVED** [review]

**Pull request merged**
